spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkAction.java

```diff
@@ -176,16 +184,55 @@ protected Dataset<Row> loadMetadataTable(Table table, MetadataTableType type) {
     return SparkTableUtil.loadMetadataTable(spark, table, type);
   }
 
-  private static class ReadManifest implements FlatMapFunction<ManifestFileBean, String> {
-    private final Broadcast<FileIO> io;
+  private static class ReadManifest implements FlatMapFunction<ManifestFileBean, ContentFileTypeBean> {
+    private final Broadcast<Table> table;
 
-    ReadManifest(Broadcast<FileIO> io) {
-      this.io = io;
+    ReadManifest(Broadcast<Table> table) {
+      this.table = table;
     }
 
     @Override
-    public Iterator<String> call(ManifestFileBean manifest) {
-      return new ClosingIterator<>(ManifestFiles.readPaths(manifest, io.getValue()).iterator());
+    public Iterator<ContentFileTypeBean> call(ManifestFileBean manifest) {
+      switch (manifest.content()) {
+        case DATA:
+          return CloseableIterator.transform(
+              ManifestFiles.read(manifest, table.getValue().io(), table.getValue().specs()).iterator(),
+              ContentFileTypeBean::toContentFilePathBean);
```

Comment:
Question: can we use a wrapper instead of creating a bean for every entry? I am a little worried about the number of generated instances and GC.

Would something like this work?

```
private CloseableIterator<ContentFileTypeBean> entries(ManifestFileBean manifest) {
  ContentFileTypeBean contentFileType = new ContentFileTypeBean();

  switch (manifest.content()) {
    case DATA:
      return CloseableIterator.transform(
          ...,
          contentFileType::wrap);
    case DELETES:
      return CloseableIterator.transform(
          ...,
          contentFileType::wrap);
    default:
      throw new IllegalArgumentException("Unsupported manifest content type:" + manifest.content());
  }
}
```

```
  public static class ContentFileTypeBean implements Serializable {
    private String file;
    private String type;

    public ContentFileTypeBean wrap(ContentFile<?> contentFile) {
      this.file = contentFile.path().toString();
      this.type = contentFile.content().toString();
      return this;
    }
  ...
}
```

Reply:
Hm a little confused how this will work, In the end, won't the iterator.next() always the same instance with same file/type which is incorrect?

Reply:
I got rid of the Bean and replaced with Scala Tuple2, makes it a bit simpler to not have the boilerplate code, but still doesn't get rid of the new instance per row.  (Still not too comfortable to replace with one single wrapper instance).